cmd/snap/cmd_run.go

```
@@ -227,6 +230,141 @@ func isReexeced() bool {
+	// having to use snap run. This code is just to ensure that a user who
+	// doesn't have those privileges can't steal the file via snap run
+	// (also note that the (potentially untrusted) snap isn't running yet).
+	fi, err := fin.Stat()
+	if err != nil {
+		return "", err
+	}
+	sys := fi.Sys()
+	if sys == nil {
+		return "", fmt.Errorf(i18n.G("Can't validate file owner"))
+	}
+	// cheap comparison as the current uid is only available as a string
+	// but it is better to convert the uid from the stat result to a
+	// string than a string into a number.
+	if fmt.Sprintf("%d", sys.(*syscall.Stat_t).Uid) != u.Uid {
+		return "", fmt.Errorf(i18n.G("Xauthority file isn't owned by the current user %s"), u.Uid)
+	}
+
+	targetPath := filepath.Join(baseTargetDir, ".Xauthority")
+
+	// Only validate Xauthority file again when both files don't match
+	// otherwise we can continue using the existing Xauthority file.
+	// This is ok to do here because we aren't trying to protect against
+	// the user changing the Xauthority file in XDG_RUNTIME_DIR outside
+	// of snapd.
+	if osutil.FileExists(targetPath) {
+		var fout *os.File
+		if fout, err = os.Open(targetPath); err != nil {
+			return "", err
+		}
+		if osutil.FileStreamsEqual(fin, fout) {
+			fout.Close()
+			return targetPath, nil
+		}
+
+		fout.Close()
+		if err := os.Remove(targetPath); err != nil {
+			logger.Noticef("WARNING: failed to remove existing file %s", targetPath)
+			return "", err
+		}
+
+		// Ensure we're validating the Xauthority file from the beginning
+		if _, err := fin.Seek(int64(os.SEEK_SET), 0); err != nil {
+			return "", err
+		}
+	}
+
+	// To guard against setting XAUTHORITY to non-xauth files, check
+	// that we have a valid Xauthority. Specifically, the file must be
+	// parseable as an Xauthority file and not be empty.
+	if err := x11.ValidateXauthorityFromFile(fin); err != nil {
+		logger.Noticef("WARNING: invalid Xauthority file: %s", err)
```

It's not clear why/when do we sometimes return the error and warn above, and sometimes warn and return nil. It'd be good to make this more consistent. If we're warning from the call site, let's return the error here as well, since this is an error similar to the other cases here, and let the call site warn too.